tls-client/HelloHttpsClient.h

```diff
@@ -0,0 +1,247 @@
+     */
+    void logTlsError( const char *func_name, int ret );
+
+    /**
+     * Wrapper function around TCPSocket that gets called by mbed TLS whenever
+     * we call mbedtls_ssl_read()
+     *
+     * \param[in]   ctx
+     *              The TCPSocket object
+     * \param[in]   buf
+     *              Buffer where data received will be stored
+     * \param[in]   len
+     *              The length (in bytes) of the buffer
+     *
+     * \return  If successful, the number of bytes received, a negative value
+     *          otherwise.
+     */
+    static int sslRecv( void *ctx, unsigned char *buf, size_t len );
+
+    /**
+     * Wrapper function around TCPSocket that gets called by mbed TLS whenever
+     * we call mbedtls_ssl_write()
+     *
+     * \param[in]   ctx
+     *              The TCPSocket object
+     * \param[in]   buf
+     *              Buffer containing the data to be sent
+     * \param[in]   len
+     *              The number of bytes to send
+     *
+     * \return  If successful, the number of bytes sent, a negative value
+     *          otherwise
+     */
+    static int sslSend( void *ctx, const unsigned char *buf, size_t len );
+
+    /**
+     * Callback to handle debug prints to serial
+     *
+     * \param[in]   ctx
+     *              The context (unused in this case)
+     * \param[in]   level
+     *              The current debug level
+     * \param[in]   file
+     *              The C file that is logging this message
+     * \param[in]   line
+     *              The line number in the file
+     * \param[in]   str
+     *              The string to log to serial
+     */
+    static void sslDebug( void *ctx, int level, const char *file, int line,
+                          const char *str );
+
+    /**
+     * Callback to handle certificate verification
+     *
+     * /param[in]       data
+     *                  (unused)
+     * /param[in]       crt
+     *                  The crt in the chain that we are verifying
+     * /param[in]       depth
+     *                  The depth of the current certificate in the chain
+     * /param[in/out]   flags
+     *                  The flags resulting from the verification
+     *
+     * /return  0 if successful
+     */
+    static int sslVerify( void *data, mbedtls_x509_crt *crt, int depth,
+                          uint32_t *flags );
+
+private:
+    /**
+     * Instance of EthernetInterface used to create a TCPSocket
+     */
+    EthernetInterface *eth_iface;
+    /**
+     * Instance of TCPSocket used to communicate with the server
+     */
+    TCPSocket *socket;
+
+    /**
+     * The domain/IP address of the server to contact
+     */
+    const char *server_name;
+    /**
+     * The port number to use in the connection
+     */
+    const uint16_t server_port;
+
+    /**
+     * A generic buffer used to hold debug or HTTP request/response strings
+     */
+    char gp_buf[GENERAL_PURPOSE_BUFFER_LENGTH];
```

Comment:
Agreed, assuming that the handshake is the peak consumption point. However, I am not sure this will cause an issue as this is allocated in the heap. What do the other reviewers think?